bidirectional LM + cnn highway encoder + gated cnn encoder

joelgrus opened the pull request:
ok, I think that this is done in an idiomatic AllenNLP way and that it addresses all of the feedback I've gotten so far.

of course this isn't the finished product, but I think these pieces are correct enough that it makes sense to merge them and start working on the next pieces.

matt-gardner commented on allennlp/modules/contextual_encoders/contextual_encoder.py:
```diff
@@ -0,0 +1,28 @@
+
+from allennlp.common.registrable import Registrable
+
+class ContextualEncoder(Registrable, torch.nn.Module):
```
I know this isn't ready for review yet; feel free to take your time answering this.  I'm just wondering what the difference is between this abstraction and our existing `Seq2SeqEncoder` abstraction.  They look identical to me.

joelgrus replied:
you're right, they're mostly identical save a property or two, let me think about how to reuse Seq2SeqEncoder here

matt-peters commented on allennlp/modules/contextual_encoder/character_encoder.py:
```diff
@@ -0,0 +1,172 @@
+from allennlp.modules.masked_layer_norm import MaskedLayerNorm
+from allennlp.nn.util import add_sentence_boundary_token_ids
+
+_DEFAULT_FILTERS = ((1, 4), (2, 8), (3, 16), (4, 32), (5, 64))
```
Make the filters required input, too much potential for hidden problem with defaults.

matt-peters commented on allennlp/tests/modules/contextual_encoder/contextual_encoder_test.py:
```diff
@@ -0,0 +1,77 @@
+        cew = CharLevelContextualEncoder(
+                encoder=self.seq2seq,
+                character_encoder=self.character_encoder,
+                num_layers=3,
```
I think `num_layers` should be 4.. it isn't a problem here, since it's never tested.
The `num_layers` thing is so that the contextual encoder can be automatically wrapped with the `Elmo` scalar weighting to make a token embedder.

joelgrus commented:
ok, I think I addressed all of your comments.

also, I just ran an evaluation on ner using this code and got the expected results:

`$ python -m allennlp.run evaluate https://s3-us-west-2.amazonaws.com/allennlp/models/ner-model-2018.04.30.tar.gz eng.testb`

```
2018-09-21 13:05:20,148 - INFO - allennlp.commands.evaluate - Finished evaluating.
2018-09-21 13:05:20,148 - INFO - allennlp.commands.evaluate - Metrics:
2018-09-21 13:05:20,148 - INFO - allennlp.commands.evaluate - accuracy: 0.9824485840422096
2018-09-21 13:05:20,148 - INFO - allennlp.commands.evaluate - accuracy3: 0.9842790998169484
2018-09-21 13:05:20,148 - INFO - allennlp.commands.evaluate - precision-overall: 0.9199789879180529
2018-09-21 13:05:20,148 - INFO - allennlp.commands.evaluate - recall-overall: 0.9302407932011332
2018-09-21 13:05:20,148 - INFO - allennlp.commands.evaluate - f1-measure-overall: 0.9250814332247058
```

matt-gardner commented on allennlp/modules/contextual_encoder/contextual_encoder.py:
```diff
@@ -0,0 +1,171 @@
+from allennlp.modules.seq2seq_encoders import Seq2SeqEncoder
+from allennlp.modules.token_embedders import TokenEmbedder
+
+class ContextualEncoder(torch.nn.Module, Registrable):
```
How does a `ContextualEncoder` get used?  Where does it get its inputs?  From a `TextField`?  I'm just worried about there being too many options for going from text to embeddings, and that the ways they are used are mutually exclusive, so you have to change how you process your data if you want to switch between them, which breaks one of the main points of the library.

Similarly, the `CharacterEncoder` looks like a fancier version of our [`CnnEncoder`](https://github.com/allenai/allennlp/blob/master/allennlp/modules/seq2vec_encoders/cnn_encoder.py) wrapped inside a `TokenCharactersEmbedder`.  It's great to have a better version of this, except now we have two ways of going through the code to get almost the same result.

It seems like we're getting a proliferation of abstractions.  I _think_ the goal of this one is to have something that wraps a `TextFieldEmbedder` and then a `Seq2SeqEncoder`, adding a few options along the way.  That's great, I see no issue there.  But the way it currently is, it doesn't actually use the `TextFieldEmbedder`, it just duplicates _almost_ the same functionality in an incompatible way, so we can't actually use things interchangeably anymore.

matt-gardner replied:
I mean, another option here is to just add a `ContextualizedTextFieldEmbedder` as a subclass of `TextFieldEmbedder`, where you group a set of `TokenEmbedders` with a `Seq2SeqEncoder`.

matt-gardner replied:
Sorry to put all of these comments here, too, but the `TokenCharactersEmbedder` already handles the mask, and the `CharacterEncoder` here doesn't actually use the mask, it just returns it.  So if you just soup up our current `CnnEncoder` with highway layers, you could remove the whole masking logic and the dictionary return type.

joelgrus replied:
no, this is good. what you are pointing out is that my approach to the Calypso -> AllenNLP project is too "literal" and that I need to step back and be a bit more holistic about it.

that is, I have been mostly approaching it as "how do I get this code working inside AllenNLP" but I should really be approaching it as "what's the right way to get this code incorporated *as part of* AllenNLP". let me think about that some.

matt-gardner replied:
Minor addendum to what I said above about the `CnnEncoder` - I wouldn't modify our existing one, as that would break our saved models.  But you could add a `CnnHighwayEncoder`, or something, that's basically your `CharacterEncoder` without the mask, and use that inside a `TokenCharactersEmbedder` to get the same effect.

joelgrus replied:
I am really struggling with the right way to do this

* TokenCharactersEncoder doesn't quite work, because it does character_ids -> embedding [add a dim] -> seq2vec_encoder [drop a dim] -> outputs

whereas ContextualEncoder wants to do character_ids -> character_encoder (which is really a seq2seq encoder) -> contextual seq2seq encoder -> outputs

--

so one possibility is to make the CharacterEncoder/CnnHighwayEncoder a TokenEmbedder, then the ContextualEncoder itself a TokenEmbedder that wraps another TokenEmbedder? (since that's really what it is)

the usage inside our current ELMo model will still be nonstandard, but then things fit more nicely into the framework?

joelgrus commented on allennlp/modules/token_embedders/cnn_highway_encoder.py:
```diff
@@ -0,0 +1,214 @@
+        'after_highway', 'after_cnn', and None.
+    do_layer_norm: bool, optional (default = False)
+        If True, we apply ``MaskedLayerNorm`` to the final encoded result.
+    bos_characters: ``List[int]``, optional (default = None)
```
I am still struggling with this a lot.

I took a step forward and implemented a bare-bones `BidirectionalLM` that looks like

```
    def __init__(self,
                 vocab: Vocabulary,
                 text_field_embedder: TextFieldEmbedder,
                 contextualizer: Seq2SeqEncoder,
                 softmax: Softmax,
                 dropout: float = None,
                 loss_scale_fac: Union[float, str] = 1.0) -> None:
```

and then has a forward that just looks like

```
    def forward(self,  # type: ignore
                tokens: Dict[str, torch.LongTensor]) -> Dict[str, torch.Tensor]:
```

in order to compute the LM targets, I insist that there was a `{"tokens": SingleIdTokenIndexer()}` and raise a `ConfigurationError` if there was not. I think this part is relatively straightforward.

I am still struggling with how to handle the rest of it.

Your suggestion (I think) was to put all of the bos/eos logic in a new `TextFieldEmbedder` subclass. But a `TextFieldEmbedder` is agnostic to the type/number of "token_indexes" it receives, whereas the bos/eos logic is specific to the actual indexes themselves? If you're doing it for characters, you have to add the bos characters; if you're doing it for tokens (do you even do it for tokens?), you just have to add a bos token, but I don't see an easy way to do it in the abstract?

matt-gardner replied:
Maybe easier to talk about this in person?

joelgrus replied:
I'm wfh today, I'll find some time tomorrow

joelgrus commented:
@matt-gardner if you could take another quick look at this and see if you agree with this approach.

basically:

* `BidirectionalLM` is a `Model`
* now `CnnHighwayEncoder` is just a `Seq2VecEncoder`, and you just use it inside the usual `TokenCharactersEncoder` with the usual `BasicTextFieldEmbedder` (it's possible there's too much overlap with the existing `CnnEncoder`, but I don't want to worry about that just yet)
* the BOS/EOS tokens get added by the `TokenIndexer`s and get removed by the model (this feels sort of clunky, but I couldn't think of a cleaner way)

I think this way is pretty allennlp-like, let me know what you think

matt-gardner commented on allennlp/data/token_indexers/single_id_token_indexer.py:
```diff
@@ -22,10 +23,23 @@ class SingleIdTokenIndexer(TokenIndexer[int]):
+    def __init__(self,
+                 namespace: str = 'tokens',
+                 lowercase_tokens: bool = False,
+                 bos_token: str = None,
```
Oh, one issue with that is that maybe the text is pre-tokenized?  Then the dataset reader would know that, and the parameters can just be added to that particular dataset reader?  I'm not sure you actually gain anything by having these parameters in this class, and it makes things easier to break.

matt-peters replied:
It seems pretty annoying for every data set reader that handles pre-tokenized text to have to worry about this particular detail to support a model that they may not use?

matt-peters replied:
Yes, in addition to training, the main use case for the `BidirectionalLanguageModel` is including a pre-trained version of one as an essential component of another supervised model as a `TextFieldEmbedder` or such.

matt-peters commented on allennlp/models/bidirectional_lm.py:
```diff
@@ -0,0 +1,180 @@
+            }
+
+        # TODO: add non-loss functions to the output
+        if self._remove_bos_eos:
```
We want to include these tokens in the loss, so that the LM learns to predict them.

matt-gardner replied:
How do you compute perplexity, then?  You strip them out before doing that?

matt-peters replied:
This code follows the convention for the 1 Billion word benchmark which for the forward LM includes `</S>` but `<S>` is masked.  The backward LM is similar but predicts `<S>` and masks `</S>`.

matt-gardner replied:
:+1:

matt-peters commented on allennlp/models/bidirectional_lm.py:
```diff
@@ -0,0 +1,180 @@
+        self._text_field_embedder = text_field_embedder
+        self._layer_norm = layer_norm
+        self._contextualizer = contextualizer
+        self._forward_dim = contextualizer.get_output_dim() // 2
```
Really we should support both bidirectional and forward only LMs with a flag or something.

matt-gardner replied:
Oh, right, the name is "bidirectional LM", so of course this has a backward component.  I guess if you want a single class to support both, you should probably rename it.  But the "LM" component just refers to the loss, right?  Whether the _contextualizer_ is bidirectional doesn't matter for this - e.g., the transformer is "directional" at all.  My point here was just about `self._forward_dim` - I was confusing myself about a bidirectional _contextualizer_, but this refers to the component that will be used for the loss.  A comment saying that explicitly would be nice.

matt-gardner replied:
It'd be nice to have a small comment explaining what this forward dim is - it's used for _predicting_ in the forward direction.

matt-gardner commented on allennlp/data/token_indexers/single_id_token_indexer.py:
```diff
@@ -43,7 +55,7 @@ def tokens_to_indices(self,
         indices: List[int] = []
 
-        for token in tokens:
+        for token in itertools.chain(self._start_tokens, tokens, self._end_tokens):
```
I need to remember to use itertools more often...  I would have just done `self._start_tokens + tokens + self._end_tokens`, which is less efficient.

matt-gardner commented on allennlp/modules/softmax.py:
```diff
@@ -0,0 +1,51 @@
+import torch
+
+class Softmax(torch.nn.Module):
+    def __init__(self,
```
This class could use a docstring describing what it's doing, and when you might want to use this instead of `torch.nn.functional.softmax`.

matt-gardner commented on allennlp/tests/models/bidirectional_lm_test.py:
```diff
@@ -0,0 +1,115 @@
+
+        self.vocab = Vocabulary.from_instances(self.instances, tokens_to_add=tokens_to_add)
+
+    def test_lm_can_run(self):
```
What does this buy you over our typical model test?  I think it's trying to test individual pieces, but those should already be tested.  It's also testing that things work if you configure them correctly, but that's already done in our typical tests (and done better, because it also checks gradients and other things).

matt-gardner commented on allennlp/tests/models/bidirectional_lm_test.py:
```diff
@@ -0,0 +1,115 @@
+
+        trainer.train()
+
+        # Loss should be lower
```
This is a piece that we could consider adding to our main model test.
